File: app/api/endpoints/audit_logs.py
"""
审计日志API端点（异步版本）

提供审计日志的查询、导出、归档、统计等功能。
"""
import os
import logging
from typing import Optional, List
from datetime import datetime
from pathlib import Path

from fastapi import APIRouter, Depends, Query, HTTPException, Request
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.responseVO import CommonResponse
from app.database import get_async_db
from app.services.audit_service import AuditLogService, get_audit_service, extract_audit_info_from_request
from app.torrents.audit_enums import AuditOperationType, AuditOperationResult
from app.auth.dependencies import get_current_user

logger = logging.getLogger(__name__)

router = APIRouter(tags=["audit-logs"])


# ========== 请求模型 ==========

class AuditLogQueryRequest(BaseModel):
    """审计日志查询请求"""
    torrent_info_id: Optional[str] = Field(None, description="种子信息ID")
    torrent_name: Optional[str] = Field(None, description="种子名称（支持模糊搜索）")
    operation_type: Optional[str] = Field(None, description="操作类型")
    operator: Optional[str] = Field(None, description="操作人")
    downloader_id: Optional[str] = Field(None, description="下载器ID")
    start_time: Optional[str] = Field(None, description="开始时间(ISO 8601格式)")
    end_time: Optional[str] = Field(None, description="结束时间(ISO 8601格式)")
    operation_result: Optional[str] = Field(None, description="操作结果")
    ip_address: Optional[str] = Field(None, description="IP地址")
    request_id: Optional[str] = Field(None, description="请求ID")
    session_id: Optional[str] = Field(None, description="会话ID")
    page: int = Field(1, ge=1, description="页码")
    page_size: int = Field(20, ge=1, le=100, description="每页数量")


class ArchiveLogsRequest(BaseModel):
    """归档审计日志请求"""
    end_time: str = Field(..., description="归档截止时间(ISO 8601格式)")
    archive_path: Optional[str] = Field(None, description="归档文件路径（可选，默认自动生成）")


class ExportLogsRequest(BaseModel):
    """导出审计日志请求"""
    torrent_info_id: Optional[str] = Field(None, description="种子信息ID")
    torrent_name: Optional[str] = Field(None, description="种子名称（支持模糊搜索）")
    operation_type: Optional[str] = Field(None, description="操作类型")
    operator: Optional[str] = Field(None, description="操作人")
    downloader_id: Optional[str] = Field(None, description="下载器ID")
    start_time: Optional[str] = Field(None, description="开始时间(ISO 8601格式)")
    end_time: Optional[str] = Field(None, description="结束时间(ISO 8601格式)")
    operation_result: Optional[str] = Field(None, description="操作结果")
    export_format: str = Field("csv", description="导出格式: csv/excel")
    max_rows: int = Field(10000, ge=1, le=100000, description="最大导出行数")


# ========== API端点 ==========

@router.post("/query", response_model=CommonResponse)
async def query_audit_logs(
    request: AuditLogQueryRequest,
    current_user = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """
    查询审计日志

    权限: 需要登录（无特殊权限限制）
    """
    try:
        # 创建审计日志服务
        audit_service = await get_audit_service(db)

        # 解析时间参数
        start_dt = datetime.fromisoformat(request.start_time) if request.start_time else None
        end_dt = datetime.fromisoformat(request.end_time) if request.end_time else None

        # 查询日志
        result = await audit_service.query_logs(
            torrent_info_id=request.torrent_info_id,
            torrent_name=request.torrent_name,  # 🔥 支持种子名称模糊搜索
            operation_type=request.operation_type,
            operator=request.operator,
            downloader_id=request.downloader_id,
            start_time=start_dt,
            end_time=end_dt,
            operation_result=request.operation_result,
            ip_address=request.ip_address,
            request_id=request.request_id,
            session_id=request.session_id,
            page=request.page,
            page_size=request.page_size
        )

        return CommonResponse(
            status="success",
            msg="查询成功",
            code="200",
            data=result
        )

    except ValueError as e:
        logger.error(f"查询审计日志失败: 参数解析错误 - {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"参数错误: {str(e)}",
            code="400",
            data=None
        )
    except Exception as e:
        logger.error(f"查询审计日志失败: {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"查询失败: {str(e)}",
            code="500",
            data=None
        )


@router.get("/statistics", response_model=CommonResponse)
async def get_audit_log_statistics(
    start_time: Optional[str] = Query(None, description="开始时间(ISO 8601格式)"),
    end_time: Optional[str] = Query(None, description="结束时间(ISO 8601格式)"),
    current_user = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """
    获取审计日志统计信息

    权限: 需要登录（无特殊权限限制）
    """
    try:
        # 创建审计日志服务
        audit_service = await get_audit_service(db)

        # 解析时间参数
        start_dt = datetime.fromisoformat(start_time) if start_time else None
        end_dt = datetime.fromisoformat(end_time) if end_time else None

        # 获取统计信息
        stats = await audit_service.get_statistics(
            start_time=start_dt,
            end_time=end_dt
        )

        return CommonResponse(
            status="success",
            msg="查询成功",
            code="200",
            data=stats
        )

    except ValueError as e:
        logger.error(f"获取审计日志统计失败: 参数解析错误 - {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"参数错误: {str(e)}",
            code="400",
            data=None
        )
    except Exception as e:
        logger.error(f"获取审计日志统计失败: {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"查询失败: {str(e)}",
            code="500",
            data=None
        )


@router.post("/archive", response_model=CommonResponse)
async def archive_audit_logs(
    request_data: ArchiveLogsRequest,
    http_request: Request,
    current_user = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """
    归档审计日志

    将指定时间之前的审计日志导出到归档文件，并从主数据库中删除。

    权限: 需要登录（无特殊权限限制）
    """
    try:
        # 创建审计日志服务
        audit_service = await get_audit_service(db)

        # 解析时间参数
        end_dt = datetime.fromisoformat(request_data.end_time)

        # 提取审计信息
        audit_info = extract_audit_info_from_request(http_request)

        # 执行归档
        result = await audit_service.archive_logs(
            end_time=end_dt,
            archive_path=request_data.archive_path
        )

        if result["success"]:
            return CommonResponse(
                status="success",
                msg=result["message"],
                code="200",
                data=result
            )
        else:
            return CommonResponse(
                status="error",
                msg=result["message"],
                code="500",
                data=result
            )

    except ValueError as e:
        logger.error(f"归档审计日志失败: 参数解析错误 - {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"参数错误: {str(e)}",
            code="400",
            data=None
        )
    except Exception as e:
        logger.error(f"归档审计日志失败: {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"归档失败: {str(e)}",
            code="500",
            data=None
        )


@router.post("/export", response_model=CommonResponse)
async def export_audit_logs(
    request_data: ExportLogsRequest,
    http_request: Request,
    current_user = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """
    导出审计日志

    支持导出为CSV或Excel格式。

    权限: 需要登录（无特殊权限限制）
    """
    try:
        # 创建审计日志服务
        audit_service = await get_audit_service(db)

        # 解析时间参数
        start_dt = datetime.fromisoformat(request_data.start_time) if request_data.start_time else None
        end_dt = datetime.fromisoformat(request_data.end_time) if request_data.end_time else None

        # 查询日志
        result = await audit_service.query_logs(
            torrent_info_id=request_data.torrent_info_id,
            torrent_name=request_data.torrent_name,  # 🔥 支持种子名称模糊搜索
            operation_type=request_data.operation_type,
            operator=request_data.operator,
            downloader_id=request_data.downloader_id,
            start_time=start_dt,
            end_time=end_dt,
            operation_result=request_data.operation_result,
            page=1,
            page_size=request_data.max_rows
        )

        if not result["list"]:
            return CommonResponse(
                status="error",
                msg="没有符合条件的数据可导出",
                code="400",
                data=None
            )

        # 导出目录
        export_dir = Path("data/audit_logs_export")
        export_dir.mkdir(parents=True, exist_ok=True)

        # 生成文件名
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        if request_data.export_format == "excel":
            output_path = export_dir / f"audit_logs_{timestamp}.xlsx"
        else:
            output_path = export_dir / f"audit_logs_{timestamp}.csv"

        # 导出：直接传递字典列表，不重建模型对象
        if request_data.export_format == "excel":
            success = await audit_service.export_logs_to_excel(result["list"], str(output_path))
        else:
            success = await audit_service.export_logs_to_csv(result["list"], str(output_path))

        if success:
            return CommonResponse(
                status="success",
                msg=f"成功导出 {len(result['list'])} 条记录",
                code="200",
                data={
                    "file_path": str(output_path),
                    "file_name": output_path.name,
                    "record_count": len(result["list"]),
                    "file_format": request_data.export_format
                }
            )
        else:
            return CommonResponse(
                status="error",
                msg="导出失败",
                code="500",
                data=None
            )

    except ValueError as e:
        logger.error(f"导出审计日志失败: 参数解析错误 - {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"参数错误: {str(e)}",
            code="400",
            data=None
        )
    except Exception as e:
        logger.error(f"导出审计日志失败: {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"导出失败: {str(e)}",
            code="500",
            data=None
        )


@router.get("/operation-types", response_model=CommonResponse)
async def get_operation_types(
    current_user = Depends(get_current_user)
):
    """
    获取所有操作类型列表

    权限: 需要登录（无特殊权限限制）
    """
    try:
        # 获取所有操作类型
        operation_types = []
        for op_type in AuditOperationType:
            operation_types.append({
                "value": op_type.value,
                "display_name": op_type.get_display_name(op_type.value),
                "category": op_type.get_category(op_type.value)
            })

        return CommonResponse(
            status="success",
            msg="查询成功",
            code="200",
            data={
                "operation_types": operation_types,
                "total": len(operation_types)
            }
        )

    except Exception as e:
        logger.error(f"获取操作类型失败: {str(e)}")
        return CommonResponse(
            status="error",
            msg=f"查询失败: {str(e)}",
            code="500",
            data=None
        )


@router.get("/download-export/{file_name}", response_model=None)
async def download_export_file(
    file_name: str,
    current_user = Depends(get_current_user)
):
    """
    下载导出的审计日志文件

    权限: 需要登录（无特殊权限限制）
    """
    from fastapi.responses import FileResponse

    try:
        # 构建文件路径
        file_path = Path("data/audit_logs_export") / file_name

        # 检查文件是否存在
        if not file_path.exists():
            raise HTTPException(status_code=404, detail="文件不存在")

        # 返回文件
        return FileResponse(
            path=str(file_path),
            filename=file_name,
            media_type='application/octet-stream'
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"下载导出文件失败: {str(e)}")
        raise HTTPException(status_code=500, detail=f"下载失败: {str(e)}")

File: frontend/src/api/audit-logs.ts
import request from '@/utils/request'
import type { ApiEnvelope } from '@/utils/request'

/**
 * 审计日志查询请求
 */
export interface AuditLogQueryRequest {
  torrent_info_id?: string
  torrent_name?: string
  operation_type?: string
  operator?: string
  downloader_id?: string
  start_time?: string
  end_time?: string
  operation_result?: string
  ip_address?: string
  request_id?: string
  session_id?: string
  page?: number
  page_size?: number
}

/**
 * 审计日志查询响应
 */
export interface AuditLogQueryResponse {
  total: number
  page: number
  pageSize: number
  list: AuditLogItem[]
}

/**
 * 审计日志项
 */
export interface AuditLogItem {
  log_id: string
  torrent_info_id: string | null
  operation_type: string
  operation_detail: string
  old_value: string | null
  new_value: string | null
  operator: string
  operation_time: string
  operation_result: string
  error_message: string | null
  downloader_id: string | null
  create_time: string
  ip_address: string | null
  user_agent: string | null
  request_id: string | null
  session_id: string | null
  torrent_name?: string | null
  downloader_name?: string | null
}

/**
 * 审计日志统计响应
 */
export interface AuditLogStatisticsResponse {
  total_count: number
  operation_type_stats: Record<string, number>
  operator_stats: Record<string, number>
  result_stats: Record<string, number>
}

/**
 * 操作类型响应
 */
export interface OperationTypesResponse {
  operation_types: OperationTypeItem[]
  total: number
}

/**
 * 操作类型项
 */
export interface OperationTypeItem {
  value: string
  display_name: string
  category: string
}

/**
 * 导出请求
 */
export interface AuditLogExportRequest {
  torrent_info_id?: string
  torrent_name?: string
  operation_type?: string
  operator?: string
  downloader_id?: string
  start_time?: string
  end_time?: string
  operation_result?: string
  export_format: 'csv' | 'excel'
  max_rows: number
}

/**
 * 导出响应
 */
export interface AuditLogExportResponse {
  file_path: string
  file_name: string
  record_count: number
  file_format: string
}

/**
 * 归档请求
 */
export interface AuditLogArchiveRequest {
  end_time: string
  archive_path?: string
}

/**
 * 归档响应
 */
export interface AuditLogArchiveResponse {
  success: boolean
  archived_count: number
  archive_path: string | null
  message: string
}

/**
 * 查询审计日志
 */
export function queryAuditLogs(params: AuditLogQueryRequest) {
  return request<ApiEnvelope<AuditLogQueryResponse>>({
    url: '/audit-logs/query',
    method: 'post',
    data: params
  })
}

/**
 * 获取审计日志统计
 */
export function getAuditLogStatistics(params?: { start_time?: string, end_time?: string }) {
  return request<ApiEnvelope<AuditLogStatisticsResponse>>({
    url: '/audit-logs/statistics',
    method: 'get',
    params
  })
}

/**
 * 获取操作类型列表
 */
export function getOperationTypes() {
  return request<ApiEnvelope<OperationTypesResponse>>({
    url: '/audit-logs/operation-types',
    method: 'get'
  })
}

/**
 * 导出审计日志
 */
export function exportAuditLogs(data: AuditLogExportRequest) {
  return request<ApiEnvelope<AuditLogExportResponse>>({
    url: '/audit-logs/export',
    method: 'post',
    data
  })
}

/**
 * 归档审计日志
 */
export function archiveAuditLogs(data: AuditLogArchiveRequest) {
  return request<ApiEnvelope<AuditLogArchiveResponse>>({
    url: '/audit-logs/archive',
    method: 'post',
    data
  })
}

/**
 * 下载导出文件（axios blob，携带认证头）
 *
 * 历史实现返回 URL 字符串供 window.open 使用，三重损坏：前缀缺 /api/v1、
 * 新标签不带 Authorization（后端 get_current_user 只认 Bearer——cookie 名
 * 与前端存储键不匹配）、完全绕过拦截器（无续期/无登出处理）。改走统一
 * request 客户端：baseURL 自动拼对前缀、请求拦截器注入 Bearer、401 走
 * 静默续期重放链路。
 */
export function downloadExportFile(fileName: string) {
  return request<Blob>({
    url: `/audit-logs/download-export/${encodeURIComponent(fileName)}`,
    method: 'get',
    responseType: 'blob'
  })
}
